src/services/jsTyping.ts

```diff
@@ -0,0 +1,282 @@
+     * @param safeList is the list of names that we are confident they are library names that requires typing
+     */
+    function getTypingNamesFromSourceFileNames(fileNames: string[]) {
+        const jsFileNames = fileNames.filter(f => ts.fileExtensionIs(f, ".js"));
+        const inferredTypingNames = jsFileNames.map(f => ts.removeFileExtension(ts.getBaseFileName(f.toLowerCase())));
+        const cleanedTypingNames = inferredTypingNames.map(f => f.replace(/((?:\.|-)min(?=\.|$))|((?:-|\.)\d+)/g, ""));
+        _safeList === undefined ? mergeTypings(cleanedTypingNames) : mergeTypings(cleanedTypingNames.filter(f => _safeList.hasOwnProperty(f)));
+    }
+
+    /**
+     * Infer typing names from node_module folder
+     * @param nodeModulesPath is the path to the "node_modules" folder
+     */
+    function getTypingNamesFromNodeModuleFolder(nodeModulesPath: string, filesToWatch: string[]) {
+        // Todo: add support for ModuleResolutionHost too
+        if (!_host.directoryExists(nodeModulesPath)) {
+            return;
+        }
+
+        const typingNames: string[] = [];
+        const packageJsonFiles =
+            _host.readDirectory(nodeModulesPath, /*extension*/undefined, /*exclude*/undefined, /*depth*/2).filter(f => ts.getBaseFileName(f) === "package.json");
+        for (const packageJsonFile of packageJsonFiles) {
+            const packageJsonDict = tryParseJson(packageJsonFile);
+            if (!packageJsonDict) { continue; }
+
+            filesToWatch.push(packageJsonFile);
+
+            // npm 3 has the package.json contains a "_requiredBy" field
+            // we should include all the top level module names for npm 2, and only module names whose
+            // "_requiredBy" field starts with "#" or equals "/" for npm 3.
+            if (packageJsonDict._requiredBy &&
+                packageJsonDict._requiredBy.filter((r: string) => r[0] === "#" || r === "/").length === 0) {
+                continue;
+            }
+            
+            // If the package has its own d.ts typings, those will take over. Otherwise the package name will be used 
+            // to download d.ts files from DefinitelyTyped
+            const packageName = packageJsonDict["name"];
+            if (packageJsonDict.hasOwnProperty("typings")) {
+                const absPath = ts.getNormalizedAbsolutePath(packageJsonDict.typings, ts.getDirectoryPath(packageJsonFile));
+                inferredTypings[packageName] = absPath;
+            }
+            else {
+                typingNames.push(packageName);
+            }
+        }
+        mergeTypings(typingNames);
+    }
+
+    function getTypingNamesFromCompilerOptions(options: CompilerOptions) {
+        const typingNames: string[] = [];
+        if (!options) {
+            return;
+        }
+
+        if (options.jsx === JsxEmit.React) {
+            typingNames.push("react");
+        }
+        if (options.moduleResolution === ModuleResolutionKind.NodeJs) {

```

Comment:
Would that be a strong enough evidence of using Node? @billti has commented before with the concern that the Common-JS style is used in many places outside the Node environment:
> Lots of folks use CommonJS-style modules and then a bundler (browserify, webpack, etc.) for web development, and aren’t necessarily doing Node development